### PR TITLE
Theming refactor and visual improvements

### LINKED ISSUE #918: New tab appears slowly, in a few visible steps
When tabs are in "large" mode, and one opens a new tab, the tab element appears in several steps over something like a second.

I’m not sure if it’s a voluntary animation with bad performance or if it looks animated — expanding vertically — because some content is shown progressively.

Things that appear after a delay:
- the "about:newtab" secundary label
- the tab preview coming late
- also the tab preview (or preview placeholder) seems to be moving around

Here is a video of the effect (real time, not slowed down):
https://dl.dropboxusercontent.com/u/145744/tabcenter-newtab-slow.mov
This is on a 2015 Macbook Pro. Note that I got the same speed on two situations: on 5% battery, and when plugged in. From that, I suspect that the issue is not affected much by the CPU speed (this computer's CPU slows down significantly at <10% battery, which affects the UI responsiveness of most applications… but here it doesn’t seem to get worse on low battery, or better when plugged in).

I was also wondering if `about:newtab` tabs really needed a preview. The placeholder image would be enough for these tabs, IMO. If that’s what's slowing down showing these tabs, that could help fix this issue too.
That would also be consistent with how the standard tab bar treats those different tabs:
- about:preferences (icon + label)
- about:addons (icon + label)
- about:newtab (label only)

### PR DESCRIPTION
**Pull request by fvsch**
Hello there,

@bwinton and @ericawright, this might not be 100% ready yet (I still need to test on Windows and Linux), but I worked on the CSS extensively to help refine the look-and-feel and the Lightweight Themes & Compact Themes compatibility of this extension.

The https://github.com/fvsch/TabCenter/tree/theming-refactor branch (and thus this PR) is a bit of a kitchen sink of different issues and small visual changes, so I’d like to present them and discuss them here.

### Included changes

(There are screenshots below.)

- Big refactor of theming using CSS Custom Properties (css vars); not all theme-related styles are handled that way, but it manages 80% of it.
- Fixes #918 (using a custom prop for the height in the animation @keyframes).
- The sidebar remains light for all themes, including "brighttext" lightweight themes, except for the "Compact Dark" theme. That was the best compromise in practice.
- For Compact themes: reuse colors defined as css vars by the theme.
- For Compact themes: make the TabsToolbar compact too, to match the navigation toolbar.
- Fixed mac theme to match the navigation bar.
- Changed the "tabs on top" icon to match the style and size of the "sidebar" icons.
- Changed the icons to be pure gray, and use CSS filters for achieving the right shade of gray or white + adding drop shadows when it matches the target theme (mostly for lightweight themes, which have white or black drop-shadows on toolbar icons).
- Tweaked the spacing of the toolbar icons, and their heights, to precisely match the layout rules used in the navigation toolbar (I used macOS as a reference, so maybe this is OS-specific? I hope not ^^).
- The "selected tab" blue thingy is now animated (parts of the CSS code hinted that was the original intent, but it wasn't working). It’s also using the Container color for container tabs, defaulting to blue.
- Selected tabs use the compact theme’s colors (white text on blue background, with 2 different shades of blue for the light and dark themes). Since it clashes with the "selected tab" blue thingy, I removed it for compact themes. I could show it for Container Tabs, though (or use the container color as background).
- New, more neutral and schematic design for the "New Tab" preview. Since the resulting PNG/b64 is much lighter, this might help with the performance of opening a new tab, too.
- I took the liberty to hide the domain (in large mode) for "about:*" documents.

### Not included

- Container Tab icons (#405). I still need to check how that information is exposed in the URL bar and if that can be mimicked here.
- Anything else related to container tabs (e.g. grouping or filtering).

### Parts I’m not happy with

- I’m detecting the compact themes with `#main-window[style*="compact.header.png"]` which is awful, because I found nothing more appropriate. If you have any info on that, or know who I could contact on this topic?
- Pretty sure Windows and Linux themes are broken at least a little bit. The theming refactor makes some rules irrelevant or contradictory. Is there a way I can test the Linux and Windows themes by staying on my current OS (macOS), rather than using VMs?

### What it looks like

theme-compact-dark
theme-compact-light
theme-lightweight-pastel
theme-lightweight-renaissance
theme-lightweight-space
theme-macos

### CODE REVIEW

**Comment by fvsch** [comment]
Build failed on CSS linting. I’ll work on fixing that asap.

**Comment by Keith94** [comment]
I like the idea of earmarking the container tabs: https://github.com/bwinton/TabCenter/issues/405#issuecomment-278489210 so the color would be visible on non-selected tabs, and be compatible with compact themes.

It would be great if you could look into implementing #854 (plus there is too much margin between tab text and close/pinned icons currently).

This PR also fixes: #838, #865, #896, #897

**Comment by fvsch** [comment]
Note to self, dark compact theme needs:
- black shadow under toolbar icons and button label
- blackish border for search field

For other themes the search field's border does not match the nav/search inputs but it's an okay approximation, and I'm not really intent on making one style per platform/version unless it looks very out of place.

**Comment by ericawright** [comment]
> Fixes #918 (using a custom prop for the height in the animation @keyframes).

YES!

>The sidebar remains light for all themes, including "brighttext" lightweight themes, except for the "Compact Dark" theme. That was the best compromise in practice.

I'm not sure why it has to be done this way? I think I'd prefer a darker sidebar in dark themes.

> New, more neutral and schematic design for the "New Tab" preview. Since the resulting PNG/b64 is much lighter, this might help with the performance of opening a new tab, too.

I like it

> Changed the "tabs on top" icon to match the style and size of the "sidebar" icons.

I don't think I like this one. If we did go with your icon I think the arrow needs to change a little bit. But also, this button's counterpart is not the "shrink" and "expand" button icons, but the "sidetabs" button icon, which appears when the tabs are put on top. It is more important that those two buttons match each other. The "sidetabs" icon is also essentially the logo for Tab Center, so I'm not sure if we'd be willing to change that one. 

>hanged the icons to be pure gray, and use CSS filters for achieving the right shade of gray or white + adding drop shadows when it matches the target theme (mostly for lightweight themes, which have white or black drop-shadows on toolbar icons).

cool

>The "selected tab" blue thingy is now animated (parts of the CSS code hinted that was the original intent, but it wasn't working). It’s also using the Container color for container tabs, defaulting to blue.

love it

> I took the liberty to hide the domain (in large mode) for "about:*" documents.

I like it, but discussed with @bwinton and we both agree that is the url bar shows the address then we should too. So this fix is great for `about:home` and `about:newtab` but all other about pages seem to show the url.

> Selected tabs use the compact theme’s colors (white text on blue background, with 2 different shades of blue for the light and dark themes). Since it clashes with the "selected tab" blue thingy, I removed it for compact themes. I could show it for Container Tabs, though (or use the container color as background).

This seems okay, I don't really see a compelling reason to make this change though. I'd want to ask for some designer feedback/opinion on this  @phlsa ? ( @phlsa see first two screenshots in the PR description)
screen shot 2017-02-21 at 11 10 02 am

- also I like how the icons change now when the browser becomes unfocused

There are some great things in here, thank you!

As for testing other OSs, nope as far as I know you can't test that within Mac OS, you'd have to use a VM or another physical machine. When I've changed styling previously I've found that Win and Linux are very different and will be 1 or 2 pixels off in many places, and different even between windows versions. I'm worried that your changes for compact themes will actually have quite a large effect in those places. Once the final design is locked in here I might be able to take over and do win/linux versions.

Note: The bug in master stopping you from making new tabs is fixed now! So you should be able to rebase and test with master

Note: I would prefer container tabs to be in a separate PR since there is already a lot in here! (if you were considering adding that)

**Comment by fvsch** [comment]
I'll keep container tabs stuff separate. For now I'm keeping the "selected tab indicator takes the container's color" tweak, but I can remove that (one line of CSS to comment).

Re. the dark themes, my rationale was:

- I don't see the lightweight themes with "bright text" setting as a dark theme per se. In my view the user intent is just to have a dark background for the top of the browser, and as a result we need light text and icons for that part. But switching the vertical tabs themselves to a dark style felt like a big assumption.
- For the Compact Dark theme, I could reuse colors defined by this theme and set as custom properties by its stylesheet. For other lightweight themes, I didn't have a design for "dark tabs", beyond what was already implemented (which had some issues).

That was just my take, I don't mind defining dark colors for the tabs. But the design spec for Tab Center doesn't have a palette for dark mode, so I'm not sure what to use. A neutral dark gray + white overlays? The blue-tinted dark grays from the Compact Dark theme? I would love @phlsa's opinion on that.

For the Compact theme's active tabs, I went the whole way of "use the theme's colors", but I'm fine with a more neutral style (darken or lighten the background).

> this button's counterpart is not the "shrink" and "expand" button icons, but the "sidetabs" button icon, which appears when the tabs are put on top

I'll restore the old design but will put it on a 16x15 grid (instead of the original 28x28 shown at 16x16) for a more consistent size.
(I could set all the icons on 16x16 but the sidebar expand/shrink icons would end up with content that is not perfectly centered. Maybe that's alright.)

> if the url bar shows the address then we should too. So this fix is great for about:home and about:newtab but all other about pages seem to show the url.

Sounds good, I'll change it for a check against `about:home`, `about:newtab` and `about:blank`.

For the record, the original design had a "Search or enter address" label (same as the URL bar), but maybe that's confusing:

**Comment by fvsch** [comment]
On the focus of theming, @ericawright, is there a reason you need specific heights for `.tabbrowser-arrowscrollbox`, such as `calc(100vh - 41px)`?
A quick test on macOS shows that `height: 100%` works just fine, but that's just one platform and on Nightly. Maybe a restriction of `height:100%` inside of a flex item in previous Fx versions ([I know it as a Chrome & Safari bug](https://github.com/philipwalton/flexbugs/issues/197), mostly).

Being able to just height:100% this element would simplify the theming since there would be fewer magic values to figure out and keep in sync.

**Comment by ericawright** [comment]
if I recall correctly that sort of thing had to do with OS differences. If you are changing that it may work, we just need to make sure to do a lot of testing on different versions.

File: .stylelintrc.json [hunk 1]
```diff
@@ -40,6 +40,25 @@
...
+        "flex-direction",
+        "flex-flow",
+        "flex-grow",
+        "flex-wrap",
+        "flex-shrink",
+        "flex-basis",
+        "justify-content",
+        "align-content",
+        "align-self",
+        "align-items",
```
Comment (bwinton) [inline .stylelintrc.json]: We don't need all these listed in the stylelint.json, and in particular, they shouldn't come before box-sizing.  Please revert this change.

File: chrome.manifest [hunk 2]
```diff
@@ -1,6 +1,6 @@
 content tabcenter vertical-tabbrowser.xml
 
 
-override  chrome://tabcenter/skin/platform.css resource://tabcenter/skin/osx.css os=Darwin
-override  chrome://tabcenter/skin/platform.css resource://tabcenter/skin/win.css os=WINNT
-override  chrome://tabcenter/skin/platform.css resource://tabcenter/skin/linux.css os=Linux
+override  chrome://tabcenter/skin/platform.css resource://tabcenter/skin/platform-mac.css os=Darwin
+override  chrome://tabcenter/skin/platform.css resource://tabcenter/skin/platform-win.css os=WINNT
+override  chrome://tabcenter/skin/platform.css resource://tabcenter/skin/platform-linux.css os=Linux
```
Comment (bwinton) [inline chrome.manifest]: I don't see what the benefit of renaming these files is.  In particular, it makes it much harder to see what changed in each of these files.  Please revert this change.

File: package.json [hunk 3]
```diff
@@ -101,6 +101,7 @@
     "changelog": "changelog",
     "build": "npm run locales && npm run lint && jpm xpi",
     "test": "npm run locales && jpm run -b nightly -p Testing",
+    "test-nocopy": "npm run locales && jpm run -b nightly -p Testing --no-copy",
```
Comment (bwinton) [inline package.json]: I don't think we need a whole new script in `package.json` for this.  Please revert this change.

File: skin/platform-linux.css [hunk 4]
```diff
@@ -1,4 +1,6 @@
-/*Only Linux Styles*/
+/**
+ * Linux style adaptations
+ */
```
Comment (bwinton) [inline skin/platform-linux.css]: I don't see why this comment needed to change.  Please revert this change (and the other two like it).

File: package.json (hunk 3 above)
Comment (fvsch) [inline package.json]: Yep, didn't meant to commit that.

File: skin/win.css [hunk 5]
```diff
@@ -1,34 +1,38 @@
-/*Only Windows Styles*/
+/**
+ * Windows style adaptations
+ */
```
Comment (bwinton) [inline skin/win.css]: (Missed one.  🙂)

**Comment by fvsch** [comment]
@bwinton On this topic I want to stress that with all the CSS variables changes plus many changes, deduplication of conflicted CSS, and finally the big reordering in https://github.com/bwinton/TabCenter/pull/944/commits/9d6b2bc5a77981f9a42c9dd78bbef457da9a4876 (in `base.css`), the resulting CSS is now so different from `base.css` on master that merges in this file are bound to fail.
I could revert that specific reorder commit, but IMO it has several benefits and even before that the CSS code had diverged a lot so it might not help.

**Comment by bwinton** [comment]
Yeah, I think those changes are good ideas, and I don't think you need to revert the whole thing.  I just meant the comments at the top.  :smile:  Thanks again, and I look forward to seeing this land once Erica reviews it, and once I test it on Windows.  (I'll be sure to post screenshots…)

**Comment by fvsch** [comment]
Here are the issues left IMO.

These two exist with the current extension:

1. Scrollbox height. I'm using variables here to make it a bit easier, but the core problem remains that in some OS/preferences/theme combinations the scrollbox should be `calc(100vh - var(--toolbar-height))` (its default value), and on others there is a fake title bar that is part of the Chrome and which adds a bit more space (22px on mac, a bit less on Win10, other values I guess on Win8, Win7). If we remove this from the scrollbox height, in settings where that fake title bar is disabled (using a separate, system or system-like toolbar instead), that means the scrollbox is too short by up to 22 pixels. I'm also not sure if the fullscreen mode with hidden nav bar adds to the issue. I haven't found ways in CSS to say "if we use the fake title bar" and "if we are in fullscreen with the nav bar hidden".
2. Fullscreen: on OSX the tab center unaffected by fullscreen (which I tend to like), on Linux and Windows it gets hidden and you have to hover/tap the left edge to see it. It's not always intuitive (e.g. on a Surface with a touch interface, it took me a long time to find a way to recall the tab center). I don't think I've broken anything on this side, though, so maybe keep that as a separate issue.

Outstanding issues with this PR:

- More tests to be done on Linux and Windows. Outside of scrollbox height, main issues seems to be button widths.
- Do we use dark colors for the lightweight "brighttext" themes, or only for the Compact Dark theme? If we want dark colors, what should they be, just plain `hsl(0, 0%, 3%)`?
- Do we keep the blue "selected tab" backgrounds for the Compact themes?
- Pretty sure I've lost the inner shadow in collapsed (non-hovered) mode. I need to fix that.
- Translucent toolbar for lightweight (non-compact) themes: nice for a pinned Tab Center, not working for the collapsed mode, so I'm using a solid color then. I'm losing the top-right corner shadow in this case, but I tend to find that better overall.
- Nice to have: change the button code to use bigger hitboxes, where the button has padding (I've seen `padding: 5px 2px;` on Windows 10, and `padding: 3px;` on Linux), and on hover the border/background are on the inner content. Not sure why the Australis theme uses this for the nav bar on Windows + Linux but not on Mac, but it could be done on all 3 platforms in Tab Center. Not a blocking issue by any means.

And any issue I haven't found. ^^

**Comment by ericawright** [comment]
re: fullscreen please don't change that for Linux and Windows. The tabs disappearing on (F!!) fullscreen was a feature requested by many users.

**Comment by fvsch** [comment]
> re: fullscreen. Please don't change that for Linux and Windows. The tabs disappearing on (F11) fullscreen was a feature requested by many users.

I don't plan to change it, mainly I'm wondering if it could be made to work on OSX as well. It doesn't seem to work for me on that OS. Is this by design?

**Comment by ericawright** [comment]
Yes, that is by design. On windows and Linux, without tab center, when the user hits F11 the tabs and urlbar etc. all disappear. This does not happen on Mac. So we are trying to emulate that behavior.

File: package.json [hunk 6]
```diff
@@ -180,4 +180,4 @@
       ]
     }
   ]
-}
\ No newline at end of file
+}
```
Comment (ericawright) [inline package.json]: ?

File: skin/base-quarantine.css [hunk 7]
```diff
@@ -0,0 +1,86 @@
+/**
```
Comment (ericawright) [inline skin/base-quarantine.css]: Fine for working, but we should not merge it with this file. Note: Please be very careful about removing things - when in doubt, just leave it as there are many different OS's and modes (tabs on top, tabs in reverse, shrunk and expanded) to test in things might surprise you.

File: package.json (hunk 6 above)
Comment (fvsch) [inline package.json]: Probably an "ensure newline at tend of file" rule from my IDE that got triggered at some point.

File: skin/base-quarantine.css (hunk 7 above)
Comment (fvsch) [inline skin/base-quarantine.css]: I'll reintegrate it in base.css, as a section at the end if things don't seem to fit in the existing sections.

**Comment by ericawright** [comment]
>  Do we use dark colors for the lightweight "brighttext" themes, or only for the Compact Dark theme? If we want dark colors, what should they be, just plain hsl(0, 0%, 3%)?
 
- yes dark colors please - the way they were worked.

>Do we keep the blue "selected tab" backgrounds for the Compact themes?

have pinged @phlsa directly about this - will get back to you

>Scrollbox height

If this is too much trouble feel free to revert it and use the measurements that were there before. We had to put quite a few hours into those calculations as it is very finicky, hard on Mac, and much much more difficult on other OSs. I know what we had was not quite perfect and there may be one or two spare pixels hanging around the way it was before as well, but we want to try to not cause new errors.

>Pretty sure I've lost the inner shadow in collapsed (non-hovered) mode. I need to fix that.
Yes please do

Icons: The size changes are okay, but only if you can keep the ratio the same - the icons look quite different. 
screen shot 2017-02-24 at 9 55 21 am
screen shot 2017-02-24 at 9 55 38 am


> More tests to be done on Linux and Windows. Outside of scrollbox height, main issues seems to be button widths.

Worry about this last I think, once everything has settled

Bugs: 
With tabs in reverse set. When using cmd-w to close a tab it selects the next tab to be the previous one, not the next one:
feb-24-2017 10-10-40

Extra pixel at the top of the tabbox:
screen shot 2017-02-24 at 10 02 09 am

Please return the blue outline on the tab search bar to match the browser's search bar
screen shot 2017-02-24 at 10 26 31 am

Resizing is very jumpy - perhaps because of the theme background?
feb-24-2017 10-35-55

**Comment by ericawright** [comment]
So I know that's a lot of stuff to work on, and you have been really fantastic with all these changes so far! Since there is so much stuff in here, and so much of it is really great I would like to get many of these changes in. However with a PR this size it is kind of hard to review and also hard to pinpoint where certain bugs might be coming from. If it is not too hard for you and only if you think it is possible, could you break some of these things into separate PR's and we could work on each thing incrementally?

**Comment by fvsch** [comment]
I agree that this PR is very large, and the CSS changes are risky.

But I’m not sure a lot can be done on the theming side without the CSS variables and some of the code reorganization (which includes fixing contradicting code, deduplication of selectors, and some housekeeping). That means leaving stuff like:
- matching compact themes (toolbar height, colors)
- matching default platform themes for the toolbar (background, icons)
- matching lightweight themes (toolbar background, but that might be at-risk for other reasons)

Basically the "ready to be merged" parts can be done without this CSS refactor, but the "almost ready" parts cannot.

**Comment by ericawright** [comment]
sounds good to me, thank you!

**Comment by fvsch** [comment]
Closing, since the current strategy is not to merge this branch.
Comments still welcome, of course.
